lib/flex_algo.h

```diff
@@ -0,0 +1,253 @@
+		return FLEX_ALGO_METRIC_TYPE_TE_DEFAULT;
+
+	return FLEX_ALGO_METRIC_TYPE_NONE;
+}
+
+/*
+ * Inline utility to convert metric-types to string.
+ */
+static inline const char *flex_algo_metric_type2str(
+	enum flex_algo_metric_type metric_type)
+{
+	switch (metric_type) {
+	case FLEX_ALGO_METRIC_TYPE_IGP:
+		return FLEX_ALGO_METRIC_TYPE_IGP_STR;
+	case FLEX_ALGO_METRIC_TYPE_UNIDIR_LINK_DELAY:
+		return FLEX_ALGO_METRIC_TYPE_UNIDIR_LINK_DELAY_STR;
+	case FLEX_ALGO_METRIC_TYPE_TE_DEFAULT:
+		return FLEX_ALGO_METRIC_TYPE_TE_DEFAULT_STR;
+	case FLEX_ALGO_METRIC_TYPE_MAX:
+	case FLEX_ALGO_METRIC_TYPE_NONE:
+		break;
+	}
+
+	return "Invalid-Metric-Type";
+}
+
+/*
+ * Defined values for Calculation-Types
+ */
+enum flex_algo_calc_type {
+	FLEX_ALGO_CALC_TYPE_MIN = 0,
+	FLEX_ALGO_CALC_TYPE_SPF = FLEX_ALGO_CALC_TYPE_MIN,	/* 0 */
+	FLEX_ALGO_CALC_TYPE_STRICT_SPF,				/* 1 */
+	FLEX_ALGO_CALC_TYPE_MAX,
+	FLEX_ALGO_CALC_TYPE_NONE
+};
+#define FLEX_ALGO_CALC_TYPE_SPF_STR "spf"
+#define FLEX_ALGO_CALC_TYPE_STRICT_SPF_STR "strict-spf"
+
+static inline enum flex_algo_calc_type flex_algo_str2calc_type(
+	const char *str)
+{
+	int len;
+
+	if (str == NULL)
+		return FLEX_ALGO_CALC_TYPE_NONE;
+
+	len = strlen(str);
+
+	if (strncmp(str, FLEX_ALGO_CALC_TYPE_SPF_STR, len) == 0)
+		return FLEX_ALGO_CALC_TYPE_SPF;
+	else if (strncmp(str, FLEX_ALGO_CALC_TYPE_STRICT_SPF_STR, len) == 0)
+		return FLEX_ALGO_CALC_TYPE_STRICT_SPF;
+
+	return FLEX_ALGO_CALC_TYPE_NONE;
+}
+
+/*
+ * Inline utility to convert metric-types to string.
+ */
+static inline const char *flex_algo_calc_type2str(
+	enum flex_algo_calc_type calc_type)
+{
+	switch (calc_type) {
+	case FLEX_ALGO_CALC_TYPE_SPF:
+		return FLEX_ALGO_CALC_TYPE_SPF_STR;
+	case FLEX_ALGO_CALC_TYPE_STRICT_SPF:
+		return FLEX_ALGO_CALC_TYPE_STRICT_SPF_STR;
+	case FLEX_ALGO_CALC_TYPE_MAX:
+	case FLEX_ALGO_CALC_TYPE_NONE:
+		break;
+	}
+
+	return "Invalid-Calculation-Type";
+}
+
+/*
+ * 4-byte AdminGroup Values
+ */
+PREDECL_LIST(flxalg_admin_groups);
+struct flex_algo_admin_grp {
+	uint32_t admin_grp;
+	struct flxalg_admin_groups_item linkage;
+};
+DECLARE_LIST(flxalg_admin_groups, struct flex_algo_admin_grp, linkage);
+#define FOREACH_FLEX_ALGO_ADMIN_GROUP(list, admingrp)			\
+	frr_each_safe(flxalg_admin_groups, (list), (admingrp))
+
+/*
+ * 4-byte SRLG Values
+ */
+PREDECL_LIST(flxalg_srlgs);
+struct flex_algo_srlg {
+	uint32_t srlg;
+	struct flxalg_srlgs_item linkage;
+};
+DECLARE_LIST(flxalg_srlgs, struct flex_algo_srlg, linkage);
+#define FOREACH_FLEX_ALGO_SRLG(list, srlg)				\
+	frr_each_safe(flxalg_srlgs, (list), (srlg))
+
+/*
+ * 4-byte Flags Values
+ */
+PREDECL_LIST(flxalg_flags);
+struct flex_algo_flags {
+	uint32_t flags;
+	struct flxalg_flags_item linkage;
+};
+DECLARE_LIST(flxalg_flags, struct flex_algo_flags, linkage);
+#define FOREACH_FLEX_ALGO_FLAGS(list, flags)				\
+	frr_each_safe(flxalg_flags, (list), (flags))
+
+/*
+ * The actual Flexible-Algorithm Definition.
+ */
+struct flex_algo_defn {
```

Comment:
Naming incorrect. This structure does not store the definition but the configuration.

Reply:
@louis-6wind I don't agree. The configuration defines the algorithm.